### PR TITLE
Implement simple backpressure for Scabbard batch submit

### PR DESCRIPTION
**Pull request by agunde406**
This PR contains several bug fixes for the following error that can occur under load
```
[2021-04-13 13:06:28.911] T["Orchestrator Outgoing"] ERROR [splinter::orchestrator::runnable]
 erminating orchestrator  outgoing thread due to error: an orchestration error occurred: 
connection 20ae95a4-b2f0-4a99-8174-29c123420c00 send queue is full
```
The above error breaks the ability to handle any further batches. 

It adds very simple back pressure to scabbard batch submit that will return TooManyRequests if 30 batches are pending. This will continue to be returned until the queue gets to half that amount. This value is arbitrary. There is not any dynamic adjustments to this number. Because only the coordinator keeps track of pending batches, two new scabbard messages were added to that will be used to tell the other services if they should be returning TooManyRequests or accepting the batches. 

The following fixes were moved to https://github.com/Cargill/splinter/pull/1309

It also logs an error and continues instead of existing if the orchestrator outgoing send queue fills up. The message is dropped if this occurs. 

Finally, the capacity of the orchestrator messages queues are increased to 512. Before they were set to 8, which is very small for message queues

### CODE REVIEW

File: services/scabbard/libscabbard/src/service/rest_api/actix/batches.rs [hunk 1]
```diff
@@ -68,6 +70,21 @@ pub fn make_add_batches_to_queue_endpoint() -> ServiceEndpoint {
                             }
                         };
 
+                        match scabbard.pending_batches() {
+                            Ok(num) => {
+                                if num >= DEFAULT_PENDING_BATCH_LIMIT {
+                                    warn!("Dropping submitted batch, too many pending batches");
```
Comment (peterschwarz) [inline services/scabbard/libscabbard/src/service/rest_api/actix/batches.rs:76]: We're not dropping it, we're rejecting the batch with the appropriate error.

**Review by peterschwarz — CHANGES_REQUESTED** [review]

File: services/scabbard/libscabbard/src/service/shared.rs [hunk 2]
```diff
@@ -114,7 +183,9 @@ impl ScabbardShared {
             );
         }
 
-        batch
+        self.update_accepting_batches()?;
```
Comment (peterschwarz) [inline services/scabbard/libscabbard/src/service/shared.rs:186]: Should this also be added to the `add_batch_to_queue` function?

Comment (agunde406) [reply services/scabbard/libscabbard/src/service/shared.rs:197]: I reworked the solution a little bit to move the updating_accepting_batches into add_batch_to_queue/pop_batch_from_queue only

**Review by peterschwarz — APPROVED** [review]

File: services/scabbard/libscabbard/src/service/shared.rs [hunk 3]
```diff
@@ -95,15 +106,84 @@ impl ScabbardShared {
...
+    fn update_accepting_batches(&mut self) -> Result<(), ScabbardError> {
+        // only the coordinator should change accepting batches and
+        // back pressure is not supported by V1
+        if !self.is_coordinator() || self.scabbard_version == ScabbardVersion::V1 {
+            return Ok(());
+        };
+
+        if self.accepting_batches {
+            // check if batch queue has grown too large
+            if self.pending_batches_count() >= DEFAULT_PENDING_BATCH_LIMIT {
```
Comment (vaporos) [inline services/scabbard/libscabbard/src/service/shared.rs:132]: Use of wrapper functions -- like pending_batches_count() -- requires the reader to reference the function unnecessarily to see the definition, then retain that definition in mind while reading the code.

File: services/scabbard/libscabbard/src/service/shared.rs [hunk 4]
```diff
@@ -95,15 +106,84 @@ impl ScabbardShared {
...
+    /// set whether we are accepting new batches
+    pub fn set_accepting_batches(&mut self, accepting: bool) {
+        self.accepting_batches = accepting;
+    }
+
+    /// Check whether the pending batch queue has either gotten too big and back pressure has
+    /// should be enabled.
+    /// If back pressure has been enabled, check to see if the queue has reduced in size enough
+    /// to start accepting batches again.
+    fn update_accepting_batches(&mut self) -> Result<(), ScabbardError> {
```
Comment (vaporos) [inline services/scabbard/libscabbard/src/service/shared.rs:123]: There are two decisions coded within this function, but those decision points occur in different points of the code, and are unrelated in terms of the algorithm. This should be refactored.

For example, this function is called when adding a batch -- but then this function checks for a change in both increasing and decreasing batch counts. When adding a batch, it is unnecessary/incorrect to check for the decreasing case.

File: services/scabbard/libscabbard/src/service/shared.rs [hunk 5]
```diff
@@ -95,15 +106,84 @@ impl ScabbardShared {
...
     }
 
-    pub fn add_batch_to_queue(&mut self, batch: BatchPair) {
+    /// Returns the number of pending batches
+    pub fn pending_batches_count(&self) -> usize {
+        self.batch_queue.len()
+    }
+
+    /// set whether we are accepting new batches
+    pub fn set_accepting_batches(&mut self, accepting: bool) {
```
Comment (vaporos) [inline services/scabbard/libscabbard/src/service/shared.rs:115]: If this does not need to be pub it should be removed and the boolean accessed directly.

File: libsplinter/src/orchestrator/mod.rs [hunk 6]
```diff
@@ -29,7 +29,7 @@ use protobuf::Message;
 
 use crate::channel;
 use crate::error::InternalError;
-use crate::mesh::{Envelope, Mesh, RecvTimeoutError as MeshRecvTimeoutError};
+use crate::mesh::{Envelope, Mesh, RecvTimeoutError as MeshRecvTimeoutError, SendError};
```
Comment (vaporos) [inline libsplinter/src/orchestrator/mod.rs:32]: This fix seems unrelated, could go in another separate PR?

File: libsplinter/src/orchestrator/builder.rs [hunk 7]
```diff
@@ -20,9 +20,9 @@ use crate::transport::Connection;
 
 use super::runnable::RunnableServiceOrchestrator;
 
-const DEFAULT_INCOMING_CAPACITY: usize = 8;
-const DEFAULT_OUTGOING_CAPACITY: usize = 8;
-const DEFAULT_CHANNEL_CAPACITY: usize = 8;
+const DEFAULT_INCOMING_CAPACITY: usize = 512;
```
Comment (vaporos) [inline libsplinter/src/orchestrator/builder.rs:23]: This commit is independent and could go into a separate PR.

**Review by vaporos — CHANGES_REQUESTED** [review]

File: services/scabbard/libscabbard/src/service/shared.rs (hunk 5 above)
Comment (agunde406) [reply services/scabbard/libscabbard/src/service/shared.rs:115]: It does need to be pub, non-coordinators use this when receiving the TOO_MANY_REQUESTS/ACCEPTING_REQUESTS message from the coordinators

File: services/scabbard/libscabbard/src/service/shared.rs [hunk 8]
```diff
@@ -95,15 +106,84 @@ impl ScabbardShared {
...
+        // only the coordinator should change accepting batches and
+        // back pressure is not supported by V1
+        if !self.is_coordinator() || self.scabbard_version == ScabbardVersion::V1 {
+            return Ok(());
+        };
+
+        if self.accepting_batches {
+            // check if batch queue has grown too large
+            if self.pending_batches_count() >= DEFAULT_PENDING_BATCH_LIMIT {
+                self.set_accepting_batches(false);
```
[the remark this answers is not in the document]
Comment (agunde406) [reply services/scabbard/libscabbard/src/service/shared.rs:133]: All of shared is wrapped in a lock.

File: services/scabbard/libscabbard/src/service/shared.rs (hunk 4 above)
Comment (agunde406) [reply services/scabbard/libscabbard/src/service/shared.rs:123]: I'll update so this code is directly in the add/pop batch queue functions instead

File: services/scabbard/libscabbard/src/service/shared.rs (hunk 3 above)
Comment (agunde406) [reply services/scabbard/libscabbard/src/service/shared.rs]: removed

File: services/scabbard/libscabbard/src/service/shared.rs (hunk 4 above)
Comment (agunde406) [reply services/scabbard/libscabbard/src/service/shared.rs]: updated, removes update_accepting_batches

File: libsplinter/src/orchestrator/mod.rs (hunk 6 above)
Comment (agunde406) [reply libsplinter/src/orchestrator/mod.rs]: moved to https://github.com/Cargill/splinter/pull/1309

File: libsplinter/src/orchestrator/builder.rs (hunk 7 above)
Comment (agunde406) [reply libsplinter/src/orchestrator/builder.rs]: moved to https://github.com/Cargill/splinter/pull/1309

**Comment by vaporos** [comment]
Because this changes the protocol, we should upgrade the scabbard version.

We should also consider upgrading the REST API protocol version because the contract has changed.

**Comment by vaporos** [comment]
This change should live behind an experimental feature so we can test it prior to stabilization.

**Comment by agunde406** [comment]
I added the experimental features. As per our discussion, the protocol version decisions will be put off until stabilization

File: services/scabbard/libscabbard/src/service/mod.rs [hunk 9]
```diff
@@ -67,7 +67,7 @@ const SERVICE_TYPE: &str = "scabbard";
 const DEFAULT_COORDINATOR_TIMEOUT: u64 = 30; // 30 seconds
 
 /// Specifies the version of scabbard to use.
-#[derive(Clone, Copy)]
+#[derive(Clone, Copy, PartialEq)]
```
Comment (peterschwarz) [inline services/scabbard/libscabbard/src/service/mod.rs:70]: Is this derive necessary? If it's only for tests, we might consider using something like

```rust
#[cfg_attr(test, derive(PartialEq))]
```

Comment (agunde406) [reply services/scabbard/libscabbard/src/service/mod.rs:70]: its required by this check https://github.com/Cargill/splinter/pull/1306/files#diff-8407c21da2f6e6d430d3275579d5723b307d0dab6cb18fda50c9edee59cf9be6R139

**Review by peterschwarz — DISMISSED** [review]

**Review by peterschwarz — APPROVED** [review]

**Review by vaporos — APPROVED** [review]
